api/index.php

```diff
@@ -83,6 +83,11 @@ function api_rate_limit($key)
         return;
     }
 
+    // memcached keys can't contain whitespaces or control characters:
+    // https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L41
+    // To enforce that, hash the key.
+    $key = md5($key);
```

Comment:
probably not a huge concern, but md5 is susceptible to [length extension attacks](https://en.wikipedia.org/wiki/Length_extension_attack) so in theory someone could create a username to cause a hash conflict on the memcache entry, causing a user to be able to block another user by making them hit their rate limit. Probably not a huge concern, but should we use a different hash algorithm? Or maybe some dumb algorithm like:
```
'user name' => 'user|name' // replace ' ' with '|'
'user|name' => 'user||name' // replace '|' with '||'
```
that should prevent any conflict on the replace unless I'm mistaken.

Reply:
Usernames are limited to 25 characters, so I'm not worried about people trying to generate hash collisions just to -- at most -- double their rate limit. I also want this to be fast since it's in the path of every API call.

Usernames are restricted to `a-zA-Z0-9@._ ` so I thought about doing a simple search/replace with some other character, but:
1. that assumes that someone wouldn't edit the list of valid characters in `pinc/User.inc` to include the now-special character
2. that the caller of this function doesn't decide to use some other key besides the username which might also contain invalid characters for memcache keys

Hashing the value seemed like the best way to make this work now and also prevent surprises by future code changes that might run into problems with other encoding solutions.

Reply:
Actually, I thought we specifically excluded the underscore character because of the wiki. Have we also now excluded things like `@` (we have quite a few usernames with that, also)?

Reply:
New registrations cannot use the underscore, but existing accounts still have them.

Reply:
See https://github.com/DistributedProofreaders/dproofreaders/blob/master/pinc/User.inc#L459 for the gory details :D